amr-wind/equation_systems/icns/icns_diffusion.H

```diff
@@ -176,24 +269,117 @@ public:
     {
         const FieldState fstate = FieldState::New;
         auto& repo = m_pdefields.repo;
+        const auto& geom = repo.mesh().Geom();
         auto& field = m_pdefields.field;
         const auto& density = m_density.state(fstate);
         const int nlevels = repo.num_active_levels();
         const int ndim = field.num_comp();
         auto rhs_ptr = repo.create_scratch_field("rhs", field.num_comp(), 0);
         const auto& viscosity = m_pdefields.mueff;
-        const auto& geom = repo.mesh().Geom();
+        const auto& mesh_fac = repo.get_field("mesh_scaling_factor_cc");
+        const auto& mesh_fac_xf = repo.get_field("mesh_scaling_factor_xf");
+        const auto& mesh_fac_yf = repo.get_field("mesh_scaling_factor_yf");
+        const auto& mesh_fac_zf = repo.get_field("mesh_scaling_factor_zf");
+
         const amrex::Real alpha = 1.0;
         const amrex::Real beta = dt;
-
         m_solver_scalar->setScalars(alpha, beta);
+
         for (int lev = 0; lev < nlevels; ++lev) {
             m_solver_scalar->setLevelBC(lev, &m_pdefields.field(lev));
-            m_solver_scalar->setACoeffs(lev, density(lev));
+        }
+
+        //////////////////////////A coeffs/////////////////////////////////
+        // TODO: Create a permanent field for det_j updated only upon regrid
+        //  scratch field for determinant
+        auto rho_times_det_j = repo.create_scratch_field(
+            1, m_density.num_grow()[0], FieldLoc::CELL);
+        for (int lev = 0; lev < nlevels; ++lev) {
+            (*rho_times_det_j)(lev).setVal(0.0);
+        }
+
+        // multiply density with detJ
+        for (int lev = 0; lev < nlevels; ++lev) {
+            for (amrex::MFIter mfi(density(lev)); mfi.isValid(); ++mfi) {
+
+                const auto& rho = density(lev).array(mfi);
+                const auto& fac = mesh_fac(lev).array(mfi);
+                const auto& prod = (*rho_times_det_j)(lev).array(mfi);
+
+                const auto& bx = mfi.growntilebox();
+                amrex::ParallelFor(
+                    bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                        amrex::Real det_j =
+                            fac(i, j, k, 0) * fac(i, j, k, 1) * fac(i, j, k, 2);
+                        prod(i, j, k, 0) = rho(i, j, k, 0) * det_j;
+                    });
+            }
+        }
+
+        // set A coeffs
+        for (int lev = 0; lev < nlevels; ++lev) {
+            m_solver_scalar->setACoeffs(lev, (*rho_times_det_j)(lev));
+        }
+        ///////////////////////////////////////////////////////////////////
+
+        //////////////////////////B coeffs/////////////////////////////////
```

Comment:
```suggestion
```

This is probably a personal preference, but I am not a big fan of unnecessary line noise.